src/hooks/sync-sites-context.tsx

```diff
@@ -0,0 +1,212 @@
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: pullStatesProgressInfo.completed,
+					downloadUrl,
+				},
+			} ) );
+
+			const filePath = await getIpcApi().downloadSyncBackup( remoteSiteId, downloadUrl );
+
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: pullStatesProgressInfo.importing,
+					downloadUrl,
+				},
+			} ) );
+
+			await importFile(
+				{
+					path: filePath,
+					type: 'application/tar+gzip',
+				},
+				selectedSite
+			);
+
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: pullStatesProgressInfo.finished,
+				},
+			} ) );
+			setTimeout( () => {
+				setPullStates( ( prevStates ) => {
+					const newStates = { ...prevStates };
+					delete newStates[ remoteSiteId ];
+					return newStates;
+				} );
+			}, 1000 );
+		},
+		[
+			importFile,
+			pullStatesProgressInfo.completed,
+			pullStatesProgressInfo.finished,
+			pullStatesProgressInfo.importing,
+		]
+	);
+
+	const getBackup = useCallback(
+		async ( remoteSiteId: number ) => {
+			if ( ! client ) {
+				return;
+			}
+			const backupId = pullStates[ remoteSiteId ]?.backupId;
+			if ( ! backupId ) {
+				console.error( 'No backup ID found' );
+				return;
+			}
+			const response = await client.req.get< {
+				status: 'in-progress' | 'completed' | 'failed';
+				download_url: string;
+			} >( `/sites/${ remoteSiteId }/studio-app/sync/backup`, {
+				apiNamespace: 'wpcom/v2',
+				backup_id: backupId,
+			} );
+
+			const statusWithProgress =
+				pullStatesProgressInfo[ response.status ] || pullStatesProgressInfo.failed;
+			const hasBackupCompleted = response.status === 'completed';
+			const downloadUrl = hasBackupCompleted ? response.download_url : null;
+
+			setPullStates( ( prevStates ) => ( {
+				...prevStates,
+				[ remoteSiteId ]: {
+					...prevStates[ remoteSiteId ],
+					status: statusWithProgress,
+					downloadUrl,
+				},
+			} ) );
+
+			if ( hasBackupCompleted && downloadUrl ) {
+				// Replacing the 'in-progress' status will stop the active listening for the backup completion
+				const selectedSite = pullStates[ remoteSiteId ]?.selectedSite;
+				if ( ! selectedSite ) {
+					return;
+				}
+				onBackupCompleted( remoteSiteId, downloadUrl, selectedSite );
+			}
+		},
+		[ client, pullStates, pullStatesProgressInfo, onBackupCompleted ]
+	);
+
+	useEffect( () => {
+		const intervals: Record< number, NodeJS.Timeout > = {};
+
+		Object.entries( pullStates ).forEach( ( [ remoteSiteId, state ] ) => {
+			if ( state.backupId && state.status.key === 'in-progress' ) {
+				intervals[ Number( remoteSiteId ) ] = setInterval( () => {
+					getBackup( Number( remoteSiteId ) );
+				}, 2000 );
+			}
+		} );
+
+		return () => {
+			Object.values( intervals ).forEach( clearInterval );
+		};
```

Comment:
Thanks for pointing out to the race condition, I changed it for setTimeout which will prevent any race condition.
I decided to add some delay between request as hundreds of users pulling production sites during hours could have some unnecessary overload in our systems, if we add a small delay it can reduce any risk and the UX will be the same. I've heard that some production sites take several hours creating the backups.